# Lucinid Mitochondrial Genomes

## Overview 
This directory contains scripts used to recover and classify lucinid bacterial symbiont MAGs for various lucinid symbiosis projects.
These scripts were used to preprocess metagenome reads, assemble scaffolds, bin MAGs, classify bacterial symbionts, and carry out functional annotation  to study their biodiversity and metabolic functions.
The code was written in BASH for Illumina metagenome reads.


## Files and Functions

| Filename                                      | Description |
|-----------------------------------------------|------------|
| `01_filter_reads.sh`                          | Filters raw sequencing reads for quality control. |
| `02_metagenomics_spades.sh`                   | Assembles metagenomic reads using SPAdes. |
| `03_metagenomics_binning.sh`                  | Performs metagenomic binning to group contigs into genomes. |
| `04_metagenomics_gtdb_classification.sh`      | Classifies metagenomic bins using GTDB. |
| `05_metagenomics_checkm2.sh`                  | Assesses bin quality using CheckM2. |
| `06_metagenomics_DRAM_function_annotation.sh` | Annotates metabolic functions using DRAM. |
| `07_metagenomics_phylogenomics.sh`            | Conducts phylogenomic analysis of metagenomic bins. |


## Prerequisites

`01_filter_reads.sh`
- bbmap

`02_metagenomics_spades.sh`
- spades

`03_metagenomics_binning.sh`
- singularity
- anaconda
- bbmap (conda)
- BWA (module)
- samtools (HPC module)
- metabat2 (singularity container)
- binsanity (singularity container)
- maxbin2 (singularity container)
- dastool (singularity container)

`04_metagenomics_gtdb_classification.sh`
- singularity
- gtdbtk v2.1.1

`05_metagenomics_checkm2.sh`


`06_metagenomics_DRAM_function_annotation.sh`


`07_metagenomics_phylogenomics.sh`


## Running the Scripts

`sbatch 01_filter_reads.sh readFile1.gz readFile2.gz`

`sbatch 02_metagenomics_spades.sh <READS.FQ> <OUTDIR>`

`sbatch 03_metagenomics_binning.sh` 
This has to be run in the directory containing the assembly scaffolds with reads in ../*fastq.gz

`sbatch 04_metagenomics_gtdb_classification.sh`
Run this in the directory containing all the MAG fasta files to be classified

`sbatch 05_metagenomics_checkm2.sh`
Run this in the directory containing all the MAG fasta files to be QCed

`sbatch 06_metagenomics_DRAM_function_annotation.sh <GENOME.fasta>`

`sbatch 07_metagenomics_phylogenomics.sh <ALIGNMENT.fasta>`
Use the alignment generated by GTDB-TK workflow. It can be found in this location (gtdb_output/align/gtdbtk.bac120.user_msa.fasta.gz) generated by 04_metagenomics_gtdb_classification.sh.

## Notes & Limitations
These scripts were designed for lucinid gill metagenome Illumina reads and for use on the GWDG HPC, and will need modifications for other data and HPCs
The scale of the projects utilizing these scripts was small and modifications are needed for scaled-up implementation. Suggest exploring snakemake or nextflow for this purpose.
I attempted to use singularity container versions of all software tools for version control and portability, however the implementation of this was imperfect and it remains mixture of conda environments and singularity containers. For better replicability and portability, I suggest using singularity containers for all tools where possible.


## Author & Contact
Written by Benedict Yuen as part of the Eco-Evolutionary Interactions group at the Max Planck Institute for Marine Microbiology. Reach me at byuen@mpi-bremen.de.
